// @bbva-web-components/bbva-spinner/bbva-spinner/bbva-spinner-styles.js
import { css } from 'lit-element';

export default css`:host {
  --bbva-spinner-primary-gradient: var(--bbva-spinner-primary-gradient-color, #2dcccd);
  --bbva-spinner-secondary-gradient: var(--bbva-spinner-secondary-gradient-color, #004481);
  --bbva-spinner-background: var(--bbva-spinner-background-color,var(--bbva-spinner-secondary-gradient));
  box-sizing: border-box;
  pointer-events: none;
  display: flex; }

:host([hidden]),
[hidden] {
  display: none !important; }

*,
*::before,
*::after {
  box-sizing: inherit; }

:host([paused]) .Filter-2,
:host([paused]) .cls-6,
:host([paused]) .Gradient,
:host([paused]) .All_Grouped {
  animation-play-state: paused; }

figure {
  margin: var(--bbva-spinner-center, auto);
  width: var(--bbva-spinner-size, 3rem);
  position: relative; }

svg {
  display: block; }

.Filter-2,
.cls-6,
.Gradient,
.All_Grouped {
  animation-duration: 1.6s;
  animation-timing-function: linear;
  animation-iteration-count: infinite;
  animation-fill-mode: both; }

.Filter-2 {
  animation-name: a1_d; }

.cls-6 {
  animation-name: Gradient_t, Gradient_fo, Gradient_f; }

.Gradient {
  animation-name: a0_t; }

.All_Grouped {
  animation-name: All_Grouped_t; }

.Gradient-0 stop:nth-child(1) {
  stop-color: var(--bbva-spinner-primary-gradient); }

.Gradient-0 stop:nth-child(2) {
  stop-color: var(--bbva-spinner-secondary-gradient); }

.Gradient-1 stop:nth-child(1) {
  stop-color: var(--bbva-spinner-primary-gradient); }

.Gradient-1 stop:nth-child(2) {
  stop-color: var(--bbva-spinner-secondary-gradient); }

.Gradient-2 stop:nth-child(1) {
  stop-color: var(--bbva-spinner-primary-gradient); }

.Gradient-2 stop:nth-child(2) {
  stop-color: var(--bbva-spinner-secondary-gradient); }

.Gradient-3 stop:nth-child(1) {
  stop-color: var(--bbva-spinner-primary-gradient); }

.Gradient-3 stop:nth-child(2) {
  stop-color: var(--bbva-spinner-primary-gradient); }

.cls-2 {
  fill: var(--bbva-spinner-background); }

@keyframes All_Grouped_t {
  0% {
    transform: translate(200px, 200px) rotate(0deg) translate(-200px, -200px);
    animation-timing-function: cubic-bezier(0.75124, 0.00386, 0.25186, 1.00696); }
  100% {
    transform: translate(200px, 200px) rotate(359.99457deg) translate(-200px, -200px); } }

@keyframes a0_t {
  0% {
    transform: scale(1, 0.7);
    animation-timing-function: cubic-bezier(0.68442, 0, 0.58, 1); }
  50% {
    transform: scale(1.4, 0.8);
    animation-timing-function: cubic-bezier(0.42, 0, 0.30322, 1); }
  100% {
    transform: scale(1, 0.7);
    animation-timing-function: cubic-bezier(0.68442, 0, 0.58, 1); } }

@keyframes Gradient_t {
  0% {
    transform: translate(0px, 0px);
    animation-timing-function: cubic-bezier(0.42, 0, 0.25, 1); }
  50% {
    transform: translate(-60px, 0px);
    animation-timing-function: cubic-bezier(0.74752, 0, 0.58, 1); }
  100% {
    transform: translate(0px, 0px); } }

@keyframes Gradient_fo {
  0% {
    fill-opacity: 0.6;
    animation-timing-function: cubic-bezier(0.42, 0, 0.58, 1); }
  50% {
    fill-opacity: 0.8;
    animation-timing-function: cubic-bezier(0.42, 0, 0.58, 1); }
  100% {
    fill-opacity: 0.6; } }

@keyframes Gradient_f {
  0% {
    fill: var(--gradient-0);
    animation-timing-function: steps(1); }
  50% {
    fill: var(--gradient-1);
    animation-timing-function: steps(1); }
  100% {
    fill: var(--gradient-2);
    animation-timing-function: steps(1); } }

@keyframes a1_d {
  0% {
    d: path("M-0.66667,0C-0.66667,0,50.5,85,167,85C281.5,85,334,0,334,0C334,0,281.5,93,167,93C50.5,93,-0.66667,0,-0.66667,0Z");
    animation-timing-function: cubic-bezier(0.87624, 0, 0.62376, 1); }
  50% {
    d: path("M-0.66667,0C-0.66667,0,41,68,167,68C290.5,68,334,0,334,0C334,0,281.425,113,166.925,113C50.4255,113,-0.66667,0,-0.66667,0Z");
    animation-timing-function: cubic-bezier(0.37376, 0, 0.12499, 1); }
  100% {
    d: path("M-0.66667,0C-0.66667,0,50.5,85,167,85C281.5,85,334,0,334,0C334,0,281.5,93,167,93C50.5,93,-0.66667,0,-0.66667,0Z");
    animation-timing-function: cubic-bezier(0.62005, 0, 0.38614, 1); } }
`;
