// @cells-components/cells-paper-scroll-header-panel/cells-paper-scroll-header-panel-styles.js
import { css } from 'lit-element';

export default css`:host {
  display: block;
  box-sizing: border-box;
  position: relative;
  overflow: hidden;
  --cells-paper-scroll-header-panel__config-margin-top: 0;
   }

:host([hidden]), [hidden] {
  display: none !important; }

*, *:before, *:after {
  box-sizing: inherit; }

#mainContainer {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  box-sizing: border-box;
  -webkit-overflow-scrolling: var(--cells-paper-scroll-header-panel-main-container-overflow-scrolling, touch);
  overflow-x: hidden;
  overflow-y: auto;
  margin-top: var(--cells-paper-scroll-header-panel__config-margin-top);
  transition: margin-top 0.2s ease-in-out;
   }

#mainContainer.configurationVisible {
  margin-top: 0; }

#headerContainer {
  position: absolute;
  top: 0;
  right: 0;
  left: 0;
  z-index: var(--cells-paper-scroll-header-panel-header-container-z-index, 1);
   }

.bg-container {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  overflow: hidden;
   }

#headerBg {
   }

#condensedHeaderBg {
   }

#headerBg,
#condensedHeaderBg {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background-repeat: no-repeat;
  background-size: cover;
  background-position: center center;
   }

#condensedHeaderBg {
  opacity: 0; }
`;
