// @bbva-web-components/bbva-header-main/bbva-header-main-styles.js
import { css } from 'lit-element';

export default css`:host {
  display: block;
  box-sizing: border-box;
  @apply --bbva-header-main; }

:host([hidden]), [hidden] {
  display: none !important; }

*, *:before, *:after {
  box-sizing: inherit; }

.header-main {
  display: flex;
  justify-content: space-between;
  align-items: center;
  position: relative;
  padding-top: var(--bbva-header-main-padding-top, 0);
  min-height: var(--bbva-header-main-min-height, calc(var(--bbva-global-spacer, 0.5rem) * 6));
  background-color: var(--bbva-header-main-bg-color, var(--bbva-navy, #072146));
  font-family: inherit;
  color: var(--bbva-header-main-color, var(--bbva-white, #FFF));
  @apply --bbva-header-main-header; }
  .header-main .text {
    overflow: hidden;
    white-space: nowrap;
    text-overflow: ellipsis;
    display: block;
    position: absolute;
    bottom: 0;
    top: var(--bbva-header-main-text-top, var(--bbva-header-main-padding-top, 0));
    left: 0;
    width: 100%;
    height: var(--bbva-header-main-min-height, calc(var(--bbva-global-spacer, 0.5rem) * 6));
    margin: 0;
    padding: 0 var(--bbva-header-main-text-padding-horizontal, calc(var(--bbva-global-spacer, 0.5rem) * 6));
    text-align: center;
    font-size: var(--bbva-header-main-text-font-size, 16px);
    line-height: var(--bbva-header-main-text-line-height, calc(var(--bbva-global-spacer, 0.5rem) * 6));
    font-weight: 400; }
    .header-main .text.full-padding {
      --bbva-header-main-text-padding-horizontal: calc(var(--bbva-global-spacer, 0.5rem) * 12); }
    .header-main .text.image {
      display: flex;
      justify-content: center; }
  .header-main .header-icons {
    display: flex; }
  .header-main .icon {
    display: block;
    position: relative;
    border: none;
    background: none;
    width: var(--bbva-header-main-min-height, calc(var(--bbva-global-spacer, 0.5rem) * 6));
    min-height: var(--bbva-header-main-min-height, calc(var(--bbva-global-spacer, 0.5rem) * 6));
    color: inherit;
    @apply --bbva-header-main-icon; }
    .header-main .icon:active, .header-main .icon:focus {
      @apply --cells-global-focus; }

/** @hostclass .modal blue background */
:host(.modal) {
  --bbva-header-main-bg-color: var(--bbva-core-blue, #004481); }

/** @hostclass .modal-three-level white background */
:host(.modal-three-level) {
  --bbva-header-main-bg-color: var(--bbva-white, #FFF);
  --bbva-header-main-color: var(--bbva-600, #121212); }

/** @hostclass .process green background */
:host(.process) {
  --bbva-header-main-bg-color: var(--bbva-dark-green, #277A3E); }

/** @hostclass .process gold background */
:host(.sign) {
  --bbva-header-main-bg-color: var(--bbva-dark-gold, #8E7022); }

/** @hostclass .transparent transparent background */
:host(.transparent) {
  --bbva-header-main-bg-color: transparent; }
`;
