// @cells-components/cells-template-paper-drawer-panel/cells-template-paper-drawer-panel/cells-template-paper-drawer-panel-styles.js
import { css } from 'lit-element';

export default css`:host {
  display: block;
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  z-index: 0;
  overflow: hidden; }

:host([state="active"]) {
  z-index: 1; }

:host([hidden]), [hidden] {
  display: none !important; }

*, *:before, *:after {
  box-sizing: inherit; }

:host([footer-fixed]) {
  --cells-paper-scroll-header-container: {
    padding-bottom: var(--cells-template-paper-drawer-panel-footer-height, 60px);

  }
; }

:host([footer-fixed]) .app__footer,
:host([has-footer]) .app__footer {
  position: absolute;
  bottom: 0;
  background-color: var(--cells-template-paper-drawer-panel-footer-bg-color, #fff);
  height: var(--cells-template-paper-drawer-panel-footer-height, 60px);
  width: 100%; }

.app-container,
.app__section {
  height: 100%; }

.app__section {
  overflow: hidden;
  position: relative;
  z-index: var(--cells-template-paper-drawer-panel-section-zindex, 0);
  background-color: var(--cells-template-paper-drawer-panel-section-bg, #fff); }

.app__header {
  z-index: var(--cells-template-paper-drawer-panel-header-zindex, 0); }

:host[relative-main] .app__main {
  position: relative; }

.app__main {
  padding-top: var(--cells-template-paper-drawer-panel-main-padding-top, 0);
  padding-bottom: var(--cells-template-paper-drawer-panel-main-padding-bottom, 0);
  width: 100%;
  height: 100%; }

/* contain the top-margins of child elements */
.app__main:before {
  content: "";
  display: table; }

:host(.overflow-initial) {
  --cells-paper-scroll-header-panel-main-container-overflow-scrolling: auto; }

.app__complementary {
  height: 100%;
  width: var(--cells-template-paper-drawer-panel-complementary-width, 100%);
  position: absolute;
  top: 0;
  bottom: 0;
  z-index: var(--cells-template-paper-drawer-panel-complementary-zindex, 0); }

.app__overlay {
  z-index: var(--cells-template-paper-drawer-panel-overlay-zindex, 0); }

.app__transactional {
  z-index: var(--cells-template-paper-drawer-panel-transactional-zindex, 0);
  position: absolute;
  bottom: 0; }
`;
